applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c

```diff
@@ -144,6 +145,72 @@ static void cap_start_worker(struct k_work *work)
 	struct le_audio_unicast_server *unicast_server =
 		&unicast_servers[idx.lvl1][idx.lvl2][idx.lvl3];
 
+	if (unicast_server->sink_ep == NULL && unicast_server->source_ep == NULL) {
+		LOG_ERR("No sink or source endpoint found for device");
+		return;
+	}
+
+	if (unicast_group_created == false) {
+		uint8_t cig_index = idx.lvl1;
+		struct bt_bap_unicast_group_stream_pair_param
+			pair_params[ARRAY_SIZE(unicast_servers[cig_index][0])];
+		/* 2 streams (one sink and one source stream) for each unicast_server */
+		struct bt_bap_unicast_group_stream_param
+			group_stream_params[(ARRAY_SIZE(unicast_servers[cig_index][0]) * 2)];
+		struct bt_bap_unicast_group_param group_param;
+
+		for (int i = 0; i < ARRAY_SIZE(group_stream_params); i++) {
+			/* Every other stream should be sink or source */
+			if ((i % 2) == 0) {
+				group_stream_params[i].qos = &lc3_preset_sink.qos;
+				group_stream_params[i].stream =
+					&unicast_servers[cig_index][0][device_iterator]
+						 .cap_sink_stream.bap_stream;
+			} else {
+				group_stream_params[i].qos = &lc3_preset_source.qos;
+				group_stream_params[i].stream =
+					&unicast_servers[cig_index][0][device_iterator]
+						 .cap_source_stream.bap_stream;
+				device_iterator++;
+			}
+		}
+
+		for (int i = 0; i < ARRAY_SIZE(pair_params); i++) {
+			if (unicast_server->sink_ep) {
+				pair_params[i].tx_param = &group_stream_params[stream_iterator];
+			} else {
+				pair_params[i].tx_param = NULL;
+			}
+			stream_iterator++;
+
+			if (unicast_server->source_ep) {
+				pair_params[i].rx_param = &group_stream_params[stream_iterator];
+			} else {
+				pair_params[i].rx_param = NULL;
+			}
+
+			stream_iterator++;
+		}
+
+		group_param.params = pair_params;
+		group_param.params_count = ARRAY_SIZE(pair_params);
+
+		if (IS_ENABLED(CONFIG_BT_AUDIO_PACKING_INTERLEAVED)) {
+			group_param.packing = BT_ISO_PACKING_INTERLEAVED;
+		} else {
+			group_param.packing = BT_ISO_PACKING_SEQUENTIAL;
+		}
+
+		ret = bt_bap_unicast_group_create(&group_param, &unicast_group);
+		if (ret) {
+			LOG_ERR("Failed to create unicast group: %d", ret);
+		} else {
+			unicast_group_created = true;
+		}
+	}
+
+	k_mutex_unlock(&mtx_unicast_group_created);
```

Comment:
I don't get the need for this mutex. It is unlocked here, hence it will always be available on line 224.

Reply:
Now the group created will happen after connected to one headset and finish the discover.
There could be a chance the unicast client connects with two headsets at the very short time, but we only need to create one unicast group. So we have to make sure the group create only happen once, and the other device needs to wait.

Reply:
I'm also wondering about this mutex. The creation of the group is protected by the boolean, so it would only guard against starting a cap procedure before the previous finished. Should the unlock then be in the unicast_start_complete_cb?

Reply:
Discuss.

Reply:
Since the bool unicast_group_created will be set after the group is created without issue, so there's a chance that two works are triggered in a very short time, and these two works are trying to create group. I found I can replicate this issue easily, so I use mutex to make sure only one work can create group.